File: flutter/signet/darwin/Classes/SignetPlugin.swift
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: 2026 Nirapod Labs

#if os(macOS)
  import FlutterMacOS
#else
  import Flutter
#endif
import Foundation
import SignetAppleCore

/// Flutter plugin entry point for the Signet Darwin binding (iOS and macOS).
///
/// Registers the generated `SignetHostApi` and forwards every call to
/// `SecureEnclaveKeyStore` in the `apple/` core. This class holds no policy and no
/// key material; the core is the only code that touches the Secure Enclave.
public class SignetPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    #if os(macOS)
      let messenger = registrar.messenger
    #else
      let messenger = registrar.messenger()
    #endif
    SignetHostApiSetup.setUp(binaryMessenger: messenger, api: SignetHostApiImpl())
  }
}

/// Bridges the generated host API to the core. Every call maps a `SignetError`
/// into a Pigeon error carrying the closed-set code; wire structs map one-to-one
/// to the core's canonical types.
final class SignetHostApiImpl: SignetHostApi {
  private let store = SecureEnclaveKeyStore()

  func generateKey(spec: KeySpecWire) throws -> GenerateResultWire {
    try coded {
      let (handle, report) = try store.generateKey(spec.toKeySpec())
      return GenerateResultWire(handleId: handle.alias, report: report.toWire())
    }
  }

  func getPublicKey(handleId: String, format: PublicKeyFormatWire) throws -> PublicKeyWire {
    try coded {
      let publicKey = try store.getPublicKey(KeyHandle(alias: handleId), format: format.toCore())
      return PublicKeyWire(
        format: publicKey.format.toWire(),
        bytes: FlutterStandardTypedData(bytes: publicKey.bytes)
      )
    }
  }

  func sign(
    handleId: String,
    digest: FlutterStandardTypedData,
    options: SignOptionsWire,
    completion: @escaping (Result<FlutterStandardTypedData, Error>) -> Void
  ) {
    do {
      let signature = try store.sign(
        KeyHandle(alias: handleId),
        digest: digest.data,
        options: options.toCore()
      )
      completion(.success(FlutterStandardTypedData(bytes: signature)))
    } catch let error as SignetError {
      completion(.failure(error.asPigeonError))
    } catch {
      completion(.failure(SignetError.hardwareError.asPigeonError))
    }
  }

  func getAttestation(handleId: String) throws -> AttestationResultWire {
    try coded { try store.getAttestation(KeyHandle(alias: handleId)).toWire() }
  }

  func getSecurityTier(handleId: String) throws -> SecurityTierReportWire {
    try coded { try store.getSecurityTier(KeyHandle(alias: handleId)).toWire() }
  }

  func exists(alias: String) throws -> Bool {
    store.exists(alias: alias)
  }

  func deleteKey(alias: String) throws {
    try coded { try store.delete(alias: alias) }
  }
}

/// Runs a core call, translating a `SignetError` into the Pigeon error carrying
/// its closed-set code. A non-`SignetError` maps to `hardwareError`. The core
/// raises only the closed set; this catch is the defensive floor, not a live path.
private func coded<T>(_ block: () throws -> T) throws -> T {
  do {
    return try block()
  } catch let error as SignetError {
    throw error.asPigeonError
  } catch let error as PigeonError {
    throw error
  } catch {
    throw SignetError.hardwareError.asPigeonError
  }
}

private extension SignetError {
  /// The closed-set wire token for this case. Pinned to `conformance/errors.json`;
  /// spelled explicitly rather than reflected, so a rename cannot drift the wire.
  var code: String {
    switch self {
    case .unavailableTier: return "unavailableTier"
    case .userCanceled: return "userCanceled"
    case .keyInvalidated: return "keyInvalidated"
    case .authFailed: return "authFailed"
    case .authContextRequired: return "authContextRequired"
    case .notFound: return "notFound"
    case .keyAlreadyExists: return "keyAlreadyExists"
    case .tierMismatchOnExisting: return "tierMismatchOnExisting"
    case .attestationUnsupported: return "attestationUnsupported"
    case .hardwareError: return "hardwareError"
    case .unsupportedPlatform: return "unsupportedPlatform"
    case .invalidArgument: return "invalidArgument"
    case .authInProgress: return "authInProgress"
    }
  }

  var asPigeonError: PigeonError {
    PigeonError(code: code, message: nil, details: nil)
  }
}

// ---- wire -> core ----

private extension KeySpecWire {
  func toKeySpec() throws -> KeySpec {
    let policy: TierPolicy
    switch tierPolicyKind {
    case .strongest:
      policy = .strongest
    case .atLeast:
      guard let atLeastClass else { throw SignetError.invalidArgument }
      policy = .atLeast(atLeastClass.toCore())
    case .bestEffort:
      policy = .bestEffort
    }
    // The Secure Enclave exposes no per-key attestation, so the wire challenge has
    // no effect on Apple; getAttestation returns format none regardless.
    return KeySpec(alias: alias, tierPolicy: policy, accessControl: .none)
  }
}

private extension HardwareClassWire {
  func toCore() -> HardwareClass {
    switch self {
    case .discreteSecure: return .discreteSecure
    case .trustedEnvironment: return .trustedEnvironment
    }
  }
}

private extension PublicKeyFormatWire {
  func toCore() -> PublicKey.Format {
    switch self {
    case .rawX962: return .rawX962
    case .spki: return .spki
    }
  }
}

private extension SignOptionsWire {
  func toCore() -> SignOptions {
    switch encoding {
    case .der: return SignOptions(encoding: .der)
    case .rawRS: return SignOptions(encoding: .rawRS)
    }
  }
}

// ---- core -> wire ----

private extension SecurityTierReport {
  func toWire() -> SecurityTierReportWire {
    let (requestedKind, requestedAtLeastClass) = requestedToWire(requested)
    return SecurityTierReportWire(
      achieved: achieved.toWire(),
      requestedKind: requestedKind,
      requestedAtLeastClass: requestedAtLeastClass,
      meetsFloor: meetsFloor,
      evidence: evidence.toWire(),
      authEnforced: authEnforced?.toWire(),
      invalidated: invalidated,
      schemaVersion: Int64(schemaVersion)
    )
  }
}

private func requestedToWire(
  _ policy: TierPolicy?
) -> (TierPolicyKindWire?, HardwareClassWire?) {
  switch policy {
  case nil: return (nil, nil)
  case .strongest?: return (.strongest, nil)
  case .atLeast(let hardwareClass)?: return (.atLeast, hardwareClass.toWire())
  case .bestEffort?: return (.bestEffort, nil)
  }
}

private extension AttestationResult {
  func toWire() -> AttestationResultWire {
    let wireFormat: AttestationFormatWire
    switch format {
    case .androidKeyChain: wireFormat = .androidKeyChain
    case .none: wireFormat = .none
    }
    let wireChain: [FlutterStandardTypedData?]? =
      chain.isEmpty ? nil : chain.map { FlutterStandardTypedData(bytes: $0) }
    return AttestationResultWire(
      format: wireFormat,
      chain: wireChain,
      schemaVersion: Int64(schemaVersion)
    )
  }
}

private extension SecurityLevel {
  func toWire() -> SecurityLevelWire {
    switch self {
    case .secureEnclave: return .secureEnclave
    case .strongBox: return .strongBox
    case .tee: return .tee
    case .tpm: return .tpm
    case .software: return .software
    }
  }
}

private extension TierEvidence {
  func toWire() -> TierEvidenceWire {
    switch self {
    case .attested: return .attested
    case .keyInfoReadback: return .keyInfoReadback
    case .seTokenPresent: return .seTokenPresent
    case .inferred: return .inferred
    case .selfReportUnverified: return .selfReportUnverified
    }
  }
}

private extension AuthClass {
  func toWire() -> AuthClassWire {
    switch self {
    case .none: return .none
    case .biometricOnly: return .biometricOnly
    case .biometricOrDeviceCredential: return .biometricOrDeviceCredential
    case .deviceCredentialOnly: return .deviceCredentialOnly
    }
  }
}

private extension HardwareClass {
  func toWire() -> HardwareClassWire {
    switch self {
    case .discreteSecure: return .discreteSecure
    case .trustedEnvironment: return .trustedEnvironment
    }
  }
}

private extension PublicKey.Format {
  func toWire() -> PublicKeyFormatWire {
    switch self {
    case .rawX962: return .rawX962
    case .spki: return .spki
    }
  }
}

File: flutter/signet/android/src/main/kotlin/org/nirapod/signet/SignetPlugin.kt
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: 2026 Nirapod Labs

package org.nirapod.signet

import androidx.fragment.app.FragmentActivity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * Flutter plugin entry point for the Signet Android binding.
 *
 * Registers the generated [SignetHostApi] and forwards every call to
 * [AndroidKeyStoreSigner] in the `android/` core. This class holds no policy and
 * no key material; the core is the only code that touches the Keystore. It is
 * [ActivityAware] so an auth-gated sign can host its biometric prompt on the
 * running activity.
 */
public class SignetPlugin : FlutterPlugin, ActivityAware {
    private var activity: FragmentActivity? = null
    private var impl: SignetHostApiImpl? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        val signer = AndroidKeyStoreSigner(binding.applicationContext)
        val hostApi = SignetHostApiImpl(signer) { activity }
        impl = hostApi
        SignetHostApi.setUp(binding.binaryMessenger, hostApi)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        SignetHostApi.setUp(binding.binaryMessenger, null)
        impl?.dispose()
        impl = null
    }

    // A biometric prompt needs a FragmentActivity to host it; the host app uses
    // FlutterFragmentActivity. A plain activity leaves this null, and a gated sign
    // then fails authContextRequired rather than crashing.
    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity as? FragmentActivity
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity as? FragmentActivity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}

/**
 * Bridges the generated host API to the core. Each non-signing call wraps the core
 * in [coded], which turns a [SignetException] into a Pigeon [FlutterError] carrying
 * the closed-set code. Signing keeps its own fail-closed mapping. Wire structs map
 * one-to-one to the core's canonical types; no policy decision is made in this
 * layer. Wire enums are the Pigeon-generated UPPER_SNAKE constants; the core enums
 * are camelCase.
 */
internal class SignetHostApiImpl(
    private val signer: AndroidKeyStoreSigner,
    private val activityProvider: () -> FragmentActivity?,
) : SignetHostApi {
    // Gated signing suspends on a main-thread biometric prompt; the core hops to
    // the main thread itself; this scope only needs to outlive the call.
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    /** Cancels any in-flight gated sign when the engine detaches. */
    fun dispose() {
        scope.cancel()
    }

    override fun generateKey(spec: KeySpecWire): GenerateResultWire = coded {
        val (handle, report) = signer.generateKey(spec.toKeySpec())
        GenerateResultWire(handleId = handle.alias, report = report.toWire())
    }

    override fun getPublicKey(handleId: String, format: PublicKeyFormatWire): PublicKeyWire = coded {
        val publicKey = signer.getPublicKey(KeyHandle(handleId), format.toCore())
        PublicKeyWire(format = publicKey.format.toWire(), bytes = publicKey.bytes)
    }

    override fun sign(
        handleId: String,
        digest: ByteArray,
        options: SignOptionsWire,
        prompt: AuthPromptWire?,
        callback: (Result<ByteArray>) -> Unit,
    ) {
        if (prompt == null) {
            // Silent path: no prompt, the synchronous core sign.
            val outcome = try {
                Result.success(signer.sign(KeyHandle(handleId), digest, options.toCore()))
            } catch (failure: Exception) {
                signFailure(failure)
            }
            callback(outcome)
            return
        }
        // Gated path: a biometric prompt needs a FragmentActivity to host it.
        val activity = activityProvider()
        if (activity == null) {
            callback(
                Result.failure(SignetException(SignetErrorCode.authContextRequired).toFlutterError()),
            )
            return
        }
        val authContext = prompt.toAuthContext(activity)
        scope.launch {
            val outcome = try {
                Result.success(signer.sign(KeyHandle(handleId), digest, authContext, options.toCore()))
            } catch (cancellation: CancellationException) {
                throw cancellation
            } catch (failure: Exception) {
                signFailure(failure)
            }
            callback(outcome)
        }
    }

    override fun getAttestation(handleId: String): AttestationResultWire = coded {
        signer.getAttestation(KeyHandle(handleId)).toWire()
    }

    override fun getSecurityTier(handleId: String): SecurityTierReportWire = coded {
        signer.getSecurityTier(KeyHandle(handleId)).toWire()
    }

    override fun exists(alias: String): Boolean = coded { signer.exists(alias) }

    override fun deleteKey(alias: String): Unit = coded { signer.delete(alias) }

    /**
     * Maps a signing failure to a closed-set [FlutterError]. A [SignetException]
     * carries its own code; anything else fails closed to `hardwareError`.
     */
    private fun signFailure(failure: Throwable): Result<ByteArray> = when (failure) {
        is SignetException -> Result.failure(failure.toFlutterError())
        else -> Result.failure(
            SignetException(SignetErrorCode.hardwareError, cause = failure).toFlutterError(),
        )
    }
}

/** Maps a [SignetException] to the Pigeon error carrying its closed-set code. */
private fun SignetException.toFlutterError(): FlutterError =
    FlutterError(code = code.name, message = message)

/** Runs a core call, translating a [SignetException] into a [FlutterError]. */
private inline fun <T> coded(block: () -> T): T =
    try {
        block()
    } catch (failure: SignetException) {
        throw failure.toFlutterError()
    }

// ---- wire -> core ----

private fun KeySpecWire.toKeySpec(): KeySpec = KeySpec(
    alias = alias,
    tierPolicy = coreTierPolicy(tierPolicyKind, atLeastClass),
    accessControl = AccessControlPolicy(
        authRequirement = authRequirement.toCore(),
        authValiditySeconds = authValiditySeconds?.toInt(),
        invalidateOnBiometricEnrollment = invalidateOnBiometricEnrollment,
    ),
    attestationChallenge = attestationChallenge,
)

private fun coreTierPolicy(
    kind: TierPolicyKindWire,
    atLeastClass: HardwareClassWire?,
): TierPolicy = when (kind) {
    TierPolicyKindWire.STRONGEST -> TierPolicy.Strongest
    TierPolicyKindWire.AT_LEAST -> TierPolicy.AtLeast(
        (atLeastClass ?: throw SignetException(SignetErrorCode.invalidArgument)).toCore(),
    )
}

private fun AuthRequirementWire.toCore(): AuthRequirement = when (this) {
    AuthRequirementWire.NONE -> AuthRequirement.none
    AuthRequirementWire.BIOMETRIC_ONLY -> AuthRequirement.biometricOnly
    AuthRequirementWire.BIOMETRIC_OR_DEVICE_CREDENTIAL -> AuthRequirement.biometricOrDeviceCredential
}

private fun AuthPromptWire.toAuthContext(activity: FragmentActivity): AuthContext = AuthContext(
    activity = activity,
    title = title,
    authRequirement = authRequirement.toCore(),
    subtitle = subtitle,
    negativeButtonText = negativeButtonText,
)

private fun HardwareClassWire.toCore(): HardwareClass = when (this) {
    HardwareClassWire.DISCRETE_SECURE -> HardwareClass.discreteSecure
    HardwareClassWire.TRUSTED_ENVIRONMENT -> HardwareClass.trustedEnvironment
}

private fun PublicKeyFormatWire.toCore(): PublicKey.Format = when (this) {
    PublicKeyFormatWire.RAW_X962 -> PublicKey.Format.rawX962
    PublicKeyFormatWire.SPKI -> PublicKey.Format.spki
}

private fun SignOptionsWire.toCore(): SignOptions = SignOptions(
    encoding = when (encoding) {
        SignEncodingWire.DER -> SignOptions.Encoding.der
        SignEncodingWire.RAW_RS -> SignOptions.Encoding.rawRS
    },
)

// ---- core -> wire ----

private fun SecurityTierReport.toWire(): SecurityTierReportWire {
    val (requestedKind, requestedAtLeastClass) = requestedToWire(requested)
    return SecurityTierReportWire(
        achieved = achieved.toWire(),
        requestedKind = requestedKind,
        requestedAtLeastClass = requestedAtLeastClass,
        evidence = evidence.toWire(),
        authEnforced = authEnforced?.toWire(),
        invalidated = invalidated,
        schemaVersion = schemaVersion.toLong(),
    )
}

private fun requestedToWire(
    policy: TierPolicy?,
): Pair<TierPolicyKindWire?, HardwareClassWire?> = when (policy) {
    null -> null to null
    TierPolicy.Strongest -> TierPolicyKindWire.STRONGEST to null
    is TierPolicy.AtLeast -> TierPolicyKindWire.AT_LEAST to policy.hardwareClass.toWire()
}

private fun AttestationResult.toWire(): AttestationResultWire = AttestationResultWire(
    format = when (format) {
        AttestationResult.Format.androidKeyChain -> AttestationFormatWire.ANDROID_KEY_CHAIN
        AttestationResult.Format.none -> AttestationFormatWire.NONE
    },
    chain = chain.ifEmpty { null },
    schemaVersion = schemaVersion.toLong(),
)

private fun SecurityLevel.toWire(): SecurityLevelWire = when (this) {
    SecurityLevel.strongBox -> SecurityLevelWire.STRONG_BOX
    SecurityLevel.tee -> SecurityLevelWire.TEE
}

private fun TierEvidence.toWire(): TierEvidenceWire = when (this) {
    TierEvidence.keyInfoReadback -> TierEvidenceWire.KEY_INFO_READBACK
}

private fun AuthClass.toWire(): AuthClassWire = when (this) {
    AuthClass.none -> AuthClassWire.NONE
    AuthClass.biometricOnly -> AuthClassWire.BIOMETRIC_ONLY
    AuthClass.biometricOrDeviceCredential -> AuthClassWire.BIOMETRIC_OR_DEVICE_CREDENTIAL
    AuthClass.deviceCredentialOnly -> AuthClassWire.DEVICE_CREDENTIAL_ONLY
}

private fun HardwareClass.toWire(): HardwareClassWire = when (this) {
    HardwareClass.discreteSecure -> HardwareClassWire.DISCRETE_SECURE
    HardwareClass.trustedEnvironment -> HardwareClassWire.TRUSTED_ENVIRONMENT
}

private fun PublicKey.Format.toWire(): PublicKeyFormatWire = when (this) {
    PublicKey.Format.rawX962 -> PublicKeyFormatWire.RAW_X962
    PublicKey.Format.spki -> PublicKeyFormatWire.SPKI
}
